libeos-payg/service.c

```diff
@@ -254,25 +262,77 @@ manager_new_cb (GObject      *source_object,
   epg_service_startup_complete (self, task, g_steal_pointer (&provider));
 }
 
+static gboolean
+clock_jump_cb (gpointer user_data)
+{
+  EpgService *self = EPG_SERVICE (user_data);
+  struct timespec ts;
+  gint64 clock_realtime_secs_v1;
+  gint64 clock_boottime_secs_v1;
+  gint64 clock_jump_delta;
+
+  if (G_UNLIKELY (clock_gettime (CLOCK_REALTIME, &ts) != 0))
+    g_error ("clock_gettime() failed for CLOCK_REALTIME: %s", g_strerror (errno));
+  clock_realtime_secs_v1 = ts.tv_sec;
+  if (G_UNLIKELY (clock_gettime (CLOCK_BOOTTIME, &ts) != 0))
+    g_error ("clock_gettime() failed for CLOCK_BOOTTIME: %s", g_strerror (errno));
+  clock_boottime_secs_v1 = ts.tv_sec;
+
+  clock_jump_delta = ((clock_realtime_secs_v1 - self->clock_realtime_secs_v0) -
+                      (clock_boottime_secs_v1 - self->clock_boottime_secs_v0));
+  if (clock_jump_delta > 0)
```

Comment:
I think you may need to allow a certain margin of error, because the pair of v1 and v0 measurements are not taken at exactly the same instant, so it's not impossible that there will be a spurious delta of 1 second (or even more, if the process happens to be descheduled at an inopportune moment.

Reply:
> What if it's negative, as it will be if CLOCK_REALTIME is set backwards in time?

You're right, it should be `if (clock_jump_delta > 0 || clock_jump_delta < 0)`

> I think you may need to allow a certain margin of error, because the pair of v1 and v0 measurements are not taken at exactly the same instant, so it's not impossible that there will be a spurious delta of 1 second (or even more, if the process happens to be descheduled at an inopportune moment.

I agree the measurement is imprecise but I don't see why we would get spurious deltas because this callback is only called if the kernel notices a discontinuous change to CLOCK_REALTIME, which I think would be unlikely to be less than a second.

Reply:
Good point. Spurious no, but the delta might be slightly wrong. I don't see a way around this and it's unlikely to be significant unless something has gone dramatically wrong with the scheduler.

Reply:
OTOH: if we're saying “this callback won't fire spuriously” then we don't need to check whether `delta` is non-zero. :-)